### PR TITLE
Fixing shellcheck warnings in push-build.sh

### PR DESCRIPTION
**Pull request by pswica**
Editing push-build.sh so that it passes shellcheck.

For context please see: https://github.com/kubernetes/release/issues/726

Specifically, the reasoning for this PR is to begin cleaning up shell scripts so that shellcheck always runs as a presubmit job.

### CODE REVIEW

**Comment by spiffxp** [comment]
/ok-to-test
/hold
until we can get a shellcheck job running against this PR

**Comment by BenTheElder** [comment]
> /override pull-release-cluster-up

uhhh @spiffxp why are we overriding this?

**Comment by BenTheElder** [comment]
(not sure what was happening, but note that this job exists because a ton of kubernetes CI picks up any changes to these scripts immediately and we had many problems with the build staging being broken by changes here, we should get that job to pass)

File: push-build.sh [hunk 1]
```diff
@@ -215,35 +214,35 @@ if [[ -n "${FLAGS_docker_registry:-}" ]]; then
...
+                                  "$RELEASE_BUCKET" "$GCS_EXTRA_PUBLISH_FILE" && break
     ((attempt++))
   done
   ((attempt>=max_attempts)) && common::exit 1 "Exiting..."
 fi
 
 # Leave push-federation-images.sh for now.  Not sure if this makes sense to
 # pull into the release repo.
-if ((FLAGS_federation)); then
+if ((${FLAGS_federation:?})); then
```
Comment (BenTheElder) [inline push-build.sh]: this change is not safe, these flags envs are dynamically created

Comment (BenTheElder) [reply push-build.sh]: (and this flag is usually not set)

**Comment by BenTheElder** [comment]
> W0628 05:48:06.312] /go/src/k8s.io/release/push-build.sh: line 239: FLAGS_federation: parameter null or not set

legitimate failure in https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/release/742/pull-release-cluster-up/1144480366258556928/

File: push-build.sh [hunk 2]
```diff
@@ -117,34 +117,33 @@ RELEASE_BUCKET=${FLAGS_bucket:-"kubernetes-release-dev"}
...
-  LATEST=$(cat $KUBE_ROOT/bazel-genfiles/version)
+  LATEST=$(cat "$KUBE_ROOT"/bazel-genfiles/version)
 else
-  LATEST=$(tar -O -xzf $KUBE_ROOT/_output/release-tars/$FLAGS_release_kind.tar.gz $FLAGS_release_kind/version)
+  LATEST=$(tar -O -xzf "$KUBE_ROOT"/_output/release-tars/$FLAGS_release_kind.tar.gz $FLAGS_release_kind/version)
 fi
-
 if [[ "$LATEST" =~ (${VER_REGEX[release]}(\.${VER_REGEX[build]})?(-dirty)?) ]]; then
-  if ((FLAGS_ci)) && [[ "$LATEST" =~ "dirty" ]]; then
+  if ((${FLAGS_ci:?})) && [[ "$LATEST" =~ "dirty" ]]; then
```
Comment (BenTheElder) [inline push-build.sh]: so all of these `FLAGS_.*` environment variables are magically generated based on flags being present, I.E> if `--ci` is passed we get `FLAGS_ci`. They're not set when the flag isn't passed.

checking them with `:?` is going to break this script.

Comment (BenTheElder) [reply push-build.sh]: I'm not sure what the best approach to dealing with these is though, the whole system is a bit magical 🤔

Comment (pswica) [reply push-build.sh]: Kubernetes is magical :)

**Comment by pswica** [comment]
@BenTheElder 
Thanks for your help with this. I think, perhaps, it may be best to sprinkle the script with shellcheck ignores when a variable needs to be piped in (I think it's better to do this than to put a shellcheck ignore at the top, in case there are legitimate instances of this error occuring). Let me know what you think?

**Comment by pswica** [comment]
Three warnings left:

```
In ./push-build.sh line 98:
source "$(dirname "$($READLINK_CMD -ne "${BASH_SOURCE[0]}")")"/lib/common.sh
       ^-- SC1090: Can't follow non-constant source. Use a directive to specify location.
 In ./push-build.sh line 99:
source "$TOOL_LIB_PATH"/gitlib.sh
       ^------------------------^ SC1090: Can't follow non-constant source. Use a directive to specify location.
 In ./push-build.sh line 100:
source "$TOOL_LIB_PATH"/releaselib.sh
       ^----------------------------^ SC1090: Can't follow non-constant source. Use a directive to specify location.
```

**Comment by pswica** [comment]
All push-build.sh warnings gone. Let me know if there's anything else I can improve on @BenTheElder

**Comment by spiffxp** [comment]
@pswica can you update `hack/.shellcheck_failures` with the other files that are currently failing?

**Comment by pswica** [comment]
> .shellcheck_failures

Sure thing:
`hack/verify-shellcheck.sh | grep "In " | awk '{ print $2 }' | grep -v SC | sort | uniq > hack/.shellcheck_failures `

**Comment by justaugustus** [comment]
@pswica -- Actually, can you remove the shellcheck failures file?
When we moved to @BenTheElder's condensed version of the script, it removed the shellcheck failures checking.

**Comment by pswica** [comment]
> @pswica -- Actually, can you remove the shellcheck failures file?
> When we moved to @BenTheElder's condensed version of the script, it removed the shellcheck failures checking.

Done!

**Comment by spiffxp** [comment]
Didn't realize we had moved to a script that doesn't use that file, so are we not keeping a list of known failures? It'd be nice to merge on green

**Comment by justaugustus** [comment]
@spiffxp -- Ideally, we just fix all the failures 😜
We shouldn't have set the job to `always_run: true` just yet.

File: push-build.sh [hunk 3]
```diff
@@ -195,15 +200,15 @@ common::stepheader COPY RELEASE ARTIFACTS
...
   else
-    release::gcs::locally_stage_release_artifacts $GCS_DEST $LATEST \
-                                                  $KUBE_ROOT/_output \
+    release::gcs::locally_stage_release_artifacts "$GCS_DEST" "$LATEST" \
+                                                  "$KUBE_ROOT"/_output \
                                                   $FLAGS_release_kind
     release::gcs::push_release_artifacts \
-     $KUBE_ROOT/_output/gcs-stage/$LATEST \
-     gs://$RELEASE_BUCKET/$GCS_DEST/$LATEST && break
+     "$KUBE_ROOT"/_output/gcs-stage/"$LATEST" \
```
Comment (spiffxp) [inline push-build.sh]: prefer quoting the whole string if it's a single argument
```suggestion
     "$KUBE_ROOT/_output/gcs-stage/$LATEST" \
```

File: push-build.sh [hunk 4]
```diff
@@ -195,15 +200,15 @@ common::stepheader COPY RELEASE ARTIFACTS
...
+     gs://"$RELEASE_BUCKET"/"$GCS_DEST"/"$LATEST" && break
```
Comment (spiffxp) [inline push-build.sh]: prefer quoting the whole string if it's a single argument
```suggestion
     "gs://$RELEASE_BUCKET/$GCS_DEST/$LATEST" && break
```

File: push-build.sh [hunk 5]
```diff
@@ -195,15 +200,15 @@ common::stepheader COPY RELEASE ARTIFACTS
...
   if $USE_BAZEL; then
-    release::gcs::bazel_push_build $GCS_DEST $LATEST $KUBE_ROOT/_output \
-                                   $RELEASE_BUCKET && break
+    release::gcs::bazel_push_build "$GCS_DEST" "$LATEST" "$KUBE_ROOT"/_output \
+                                   "$RELEASE_BUCKET" && break
   else
-    release::gcs::locally_stage_release_artifacts $GCS_DEST $LATEST \
-                                                  $KUBE_ROOT/_output \
+    release::gcs::locally_stage_release_artifacts "$GCS_DEST" "$LATEST" \
+                                                  "$KUBE_ROOT"/_output \
```
Comment (spiffxp) [inline push-build.sh]:
```suggestion
                                                  "$KUBE_ROOT/_output" \
```

File: push-build.sh [hunk 6]
```diff
@@ -215,35 +220,36 @@ if [[ -n "${FLAGS_docker_registry:-}" ]]; then
   ##############################################################################
   # TODO: support Bazel too
   # Docker tags cannot contain '+'
-  release::docker::release $FLAGS_docker_registry ${LATEST/+/_} \
-    $KUBE_ROOT/_output
+  release::docker::release "$FLAGS_docker_registry" "${LATEST/+/_}" \
+    "$KUBE_ROOT"/_output
```
Comment (spiffxp) [inline push-build.sh]:
```suggestion
    "$KUBE_ROOT/_output"
```

File: push-build.sh [hunk 7]
```diff
@@ -195,15 +200,15 @@ common::stepheader COPY RELEASE ARTIFACTS
 attempt=0
 while ((attempt<max_attempts)); do
   if $USE_BAZEL; then
-    release::gcs::bazel_push_build $GCS_DEST $LATEST $KUBE_ROOT/_output \
-                                   $RELEASE_BUCKET && break
+    release::gcs::bazel_push_build "$GCS_DEST" "$LATEST" "$KUBE_ROOT"/_output \
```
Comment (spiffxp) [inline push-build.sh]:
```suggestion
    release::gcs::bazel_push_build "$GCS_DEST" "$LATEST" "$KUBE_ROOT/_output" \
```

File: push-build.sh [hunk 8]
```diff
@@ -215,35 +220,36 @@ if [[ -n "${FLAGS_docker_registry:-}" ]]; then
...
 if ! ((FLAGS_noupdatelatest)); then
   ##############################################################################
-  common::stepheader UPLOAD to $RELEASE_BUCKET
+  common::stepheader UPLOAD to "$RELEASE_BUCKET"
   ##############################################################################
   attempt=0
   while ((attempt<max_attempts)); do
-    release::gcs::publish_version $GCS_DEST $LATEST $KUBE_ROOT/_output \
-                                  $RELEASE_BUCKET $GCS_EXTRA_PUBLISH_FILE && break
+    release::gcs::publish_version "$GCS_DEST" "$LATEST" "$KUBE_ROOT"/_output \
```
Comment (spiffxp) [inline push-build.sh]:
```suggestion
    release::gcs::publish_version "$GCS_DEST" "$LATEST" "$KUBE_ROOT/_output" \
```

File: push-build.sh [hunk 9]
```diff
@@ -215,35 +220,36 @@ if [[ -n "${FLAGS_docker_registry:-}" ]]; then
...
+                                  "$RELEASE_BUCKET" "$GCS_EXTRA_PUBLISH_FILE" && break
     ((attempt++))
   done
   ((attempt>=max_attempts)) && common::exit 1 "Exiting..."
 fi
 
 # Leave push-federation-images.sh for now.  Not sure if this makes sense to
 # pull into the release repo.
+# shellcheck disable=SC2154
```
Comment (spiffxp) [inline push-build.sh]: Can you add a comment explaining why this is necessary?

File: push-build.sh [hunk 10]
```diff
@@ -215,35 +220,36 @@ if [[ -n "${FLAGS_docker_registry:-}" ]]; then
...
 if ((FLAGS_federation)); then
   ############################################################################
   common::stepheader PUSH FEDERATION
   ############################################################################
   logecho -n "Push federation images: "
   # FEDERATION_PUSH_REPO_BASE should be set by the calling job (yaml)
-  logrun -s ${KUBE_ROOT}/federation/develop/push-federation-images.sh
+  logrun -s "${KUBE_ROOT}"/federation/develop/push-federation-images.sh
```
Comment (spiffxp) [inline push-build.sh]:
```suggestion
  logrun -s "${KUBE_ROOT}/federation/develop/push-federation-images.sh"
```

File: push-build.sh [hunk 11]
```diff
@@ -117,34 +120,34 @@ RELEASE_BUCKET=${FLAGS_bucket:-"kubernetes-release-dev"}
...
 USE_BAZEL=false
-if release::was_built_with_bazel $KUBE_ROOT $FLAGS_release_kind; then
+if release::was_built_with_bazel "$KUBE_ROOT" $FLAGS_release_kind; then
   USE_BAZEL=true
   bazel build //:version
-  LATEST=$(cat $KUBE_ROOT/bazel-genfiles/version)
+  LATEST=$(cat "$KUBE_ROOT"/bazel-genfiles/version)
 else
-  LATEST=$(tar -O -xzf $KUBE_ROOT/_output/release-tars/$FLAGS_release_kind.tar.gz $FLAGS_release_kind/version)
+  LATEST=$(tar -O -xzf "$KUBE_ROOT"/_output/release-tars/$FLAGS_release_kind.tar.gz $FLAGS_release_kind/version)
```
Comment (spiffxp) [inline push-build.sh]:
```suggestion
  LATEST=$(tar -O -xzf "$KUBE_ROOT/_output/release-tars/$FLAGS_release_kind.tar.gz" $FLAGS_release_kind/version)
```

File: push-build.sh [hunk 12]
```diff
@@ -95,16 +95,19 @@ else
   READLINK_CMD=readlink
 fi
 
-source $(dirname $($READLINK_CMD -ne $BASH_SOURCE))/lib/common.sh
-source $TOOL_LIB_PATH/gitlib.sh
-source $TOOL_LIB_PATH/releaselib.sh
+# shellcheck source=./lib/common.sh
+source "$(dirname "$($READLINK_CMD -ne "${BASH_SOURCE[0]}")")"/lib/common.sh
+# shellcheck source=./lib/gitlib.sh
+source "$TOOL_LIB_PATH"/gitlib.sh
```
Comment (spiffxp) [inline push-build.sh]:
```suggestion
source "$TOOL_LIB_PATH/gitlib.sh"
```

File: push-build.sh [hunk 13]
```diff
@@ -95,16 +95,19 @@ else
...
+# shellcheck source=./lib/releaselib.sh
+source "$TOOL_LIB_PATH"/releaselib.sh
```
Comment (spiffxp) [inline push-build.sh]:
```suggestion
source "$TOOL_LIB_PATH/releaselib.sh"
```

File: push-build.sh [hunk 14]
```diff
@@ -117,34 +120,34 @@ RELEASE_BUCKET=${FLAGS_bucket:-"kubernetes-release-dev"}
...
 # This will canonicalize the path
 KUBE_ROOT=$(pwd -P)
 
 USE_BAZEL=false
-if release::was_built_with_bazel $KUBE_ROOT $FLAGS_release_kind; then
+if release::was_built_with_bazel "$KUBE_ROOT" $FLAGS_release_kind; then
   USE_BAZEL=true
   bazel build //:version
-  LATEST=$(cat $KUBE_ROOT/bazel-genfiles/version)
+  LATEST=$(cat "$KUBE_ROOT"/bazel-genfiles/version)
```
Comment (spiffxp) [inline push-build.sh]:
```suggestion
  LATEST=$(cat "$KUBE_ROOT/bazel-genfiles/version")
```

File: push-build.sh [hunk 15]
```diff
@@ -117,34 +120,34 @@ RELEASE_BUCKET=${FLAGS_bucket:-"kubernetes-release-dev"}
...
 else
-  LATEST=$(tar -O -xzf $KUBE_ROOT/_output/release-tars/$FLAGS_release_kind.tar.gz $FLAGS_release_kind/version)
+  LATEST=$(tar -O -xzf "$KUBE_ROOT"/_output/release-tars/$FLAGS_release_kind.tar.gz $FLAGS_release_kind/version)
 fi
-
 if [[ "$LATEST" =~ (${VER_REGEX[release]}(\.${VER_REGEX[build]})?(-dirty)?) ]]; then
+  # shellcheck disable=SC2154
```
Comment (spiffxp) [inline push-build.sh]: Can we explain why the disable is needed? Taking a guess:
```suggestion
  # FLAGS_ vars are set dynamically by common::namevalue(); shellcheck can't see their assignments, and we are explicitly checking if they exist
  # shellcheck disable=SC2154
```

File: push-build.sh [hunk 16]
```diff
@@ -155,9 +158,11 @@ fi
 GCS_DEST="devel"
 ((FLAGS_ci)) && GCS_DEST="ci"
 GCS_DEST=${FLAGS_release_type:-$GCS_DEST}
-GCS_DEST+="$FLAGS_gcs_suffix"
+# shellcheck disable=SC2154
```
Comment (spiffxp) [inline push-build.sh]: Can we explain why the disable is needed?

Comment (pswica) [reply push-build.sh]: I'm disabling any variable that gets declared outside of the script. I couldn't really think of a better way (when I defined the variable as ${VARIABLE_NAME:?}, for example, it broke the script on test and it was found to be dangerous.

File: push-build.sh (hunk 15 above)
Comment (pswica) [reply push-build.sh]: Your guess is correct :)

File: push-build.sh (hunk 16 above)
Comment (spiffxp) [reply push-build.sh]: sorry, I mean in a comment in the file, it's nicer for maintainers down the line to understand why these are here

Comment (pswica) [reply push-build.sh]: Oh I see. For sure, will do

File: push-build.sh (hunk 15 above)
Comment (pswica) [reply push-build.sh]: Will explain in a comment

**Comment by justaugustus** [comment]
Chatted with @spiffxp and I realized I misinterpreted the test-infra job config. 
We can definitely merge this once the push-build stuff is fixed up.

**Comment by pswica** [comment]
@justaugustus @spiffxp 
Thanks guys :) I think I got all the changes done. This helped because I'm gonna be doing a lot of these cleanups as a long-weekend project. Main lesson: `"$var/somedata"` not `"$var"/somedata`!

Hopefully we won't be in the state of constant pull-release-failures for long

**Comment by justaugustus** [comment]
Thanks so much for working on this, @pswica!
/lgtm
/approve

**Pull request merged**